File: backend/src/main/java/com/hostel/backend/service/ComplaintRatingService.java
package com.hostel.backend.service;

import com.hostel.backend.dto.RatingRequest;
import com.hostel.backend.dto.RatingResponse;
import com.hostel.backend.dto.WorkerRatingSummary;
import com.hostel.backend.entity.Complaint;
import com.hostel.backend.entity.ComplaintRating;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.ComplaintStatus;
import com.hostel.backend.enums.Role;
import com.hostel.backend.exception.ResourceNotFoundException;
import com.hostel.backend.exception.UnauthorizedException;
import com.hostel.backend.repository.ComplaintRatingRepository;
import com.hostel.backend.repository.ComplaintRepository;
import com.hostel.backend.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ComplaintRatingService {

    private final ComplaintRatingRepository ratingRepository;
    private final ComplaintRepository complaintRepository;
    private final UserRepository userRepository;

    public ComplaintRatingService(ComplaintRatingRepository ratingRepository,
                                   ComplaintRepository complaintRepository,
                                   UserRepository userRepository) {
        this.ratingRepository    = ratingRepository;
        this.complaintRepository = complaintRepository;
        this.userRepository      = userRepository;
    }

    @Transactional
    public RatingResponse submitRating(Long complaintId, String email, RatingRequest request) {
        User student = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        if (student.getRole() != Role.STUDENT) {
            throw new UnauthorizedException("Only students can rate complaints");
        }
        if (request.getRating() < 1 || request.getRating() > 5) {
            throw new RuntimeException("Rating must be between 1 and 5");
        }
        Complaint complaint = complaintRepository.findById(complaintId)
                .orElseThrow(() -> new ResourceNotFoundException("Complaint not found"));
        if (!complaint.getStudent().getId().equals(student.getId())) {
            throw new UnauthorizedException("You can only rate your own complaints");
        }
        if (complaint.getStatus() != ComplaintStatus.CLOSED) {
            throw new RuntimeException("You can only rate closed complaints");
        }
        if (complaint.getAssignedWorker() == null) {
            throw new RuntimeException("No worker assigned to this complaint");
        }
        if (ratingRepository.existsByComplaintId(complaintId)) {
            throw new RuntimeException("You have already rated this complaint");
        }
        ComplaintRating rating = ComplaintRating.builder()
                .complaintId(complaintId)
                .workerId(complaint.getAssignedWorker().getId())
                .studentId(student.getId())
                .rating(request.getRating())
                .comment(request.getComment())
                .build();
        return toResponse(ratingRepository.save(rating),
                complaint.getAssignedWorker().getName());
    }

    @Transactional(readOnly = true)
    public RatingResponse getRatingForComplaint(Long complaintId) {
        ComplaintRating rating = ratingRepository.findByComplaintId(complaintId)
                .orElse(null);
        if (rating == null) return null;
        User worker = userRepository.findById(rating.getWorkerId()).orElse(null);
        return toResponse(rating, worker != null ? worker.getName() : "Unknown");
    }

    @Transactional(readOnly = true)
    public WorkerRatingSummary getWorkerRatingSummary(Long workerId) {
        User worker = userRepository.findById(workerId)
                .orElseThrow(() -> new ResourceNotFoundException("Worker not found"));
        Double avg   = ratingRepository.getAverageRatingByWorkerId(workerId);
        Long   count = ratingRepository.getRatingCountByWorkerId(workerId);
        return WorkerRatingSummary.builder()
                .workerId(workerId)
                .workerName(worker.getName())
                .department(worker.getDepartment() != null
                        ? worker.getDepartment().getName() : null) // ✅ fixed
                .averageRating(avg != null ? Math.round(avg * 10.0) / 10.0 : 0.0)
                .totalRatings(count != null ? count : 0L)
                .build();
    }

    @Transactional(readOnly = true)
    public List<WorkerRatingSummary> getAllWorkerRatings() {
        return userRepository.findByRole(Role.WORKER)
                .stream()
                .map(w -> getWorkerRatingSummary(w.getId()))
                .collect(Collectors.toList());
    }

    private RatingResponse toResponse(ComplaintRating r, String workerName) {
        return RatingResponse.builder()
                .id(r.getId())
                .complaintId(r.getComplaintId())
                .workerId(r.getWorkerId())
                .workerName(workerName)
                .rating(r.getRating())
                .comment(r.getComment())
                .createdAt(r.getCreatedAt())
                .build();
    }
}

File: backend/src/services/complaintRating.service.js
const prisma = require('../config/prisma');
const { NotFoundError, ForbiddenError, BadRequestError } = require('../utils/AppError');

// Mirrors ComplaintRatingService.java. NOTE: this is the standalone
// `complaint_ratings` table / rating flow — separate from (and in addition
// to) the star rating stored directly on `complaints.rating` via the
// close-complaint flow in complaint.service.js. Both exist in the original
// app; neither one is removed here.

function toResponse(r, workerName) {
  return {
    id: r.id,
    complaintId: r.complaintId,
    workerId: r.workerId,
    workerName,
    rating: r.rating,
    comment: r.comment,
    createdAt: r.createdAt,
  };
}

async function submitRating(complaintId, email, request) {
  const student = await prisma.user.findUnique({ where: { email } });
  if (!student) throw new NotFoundError('User not found');
  if (student.role !== 'STUDENT') throw new ForbiddenError('Only students can rate complaints');
  if (request.rating < 1 || request.rating > 5) throw new BadRequestError('Rating must be between 1 and 5');

  const complaint = await prisma.complaint.findUnique({ where: { id: BigInt(complaintId) } });
  if (!complaint) throw new NotFoundError('Complaint not found');
  if (complaint.studentId !== student.id) throw new ForbiddenError('You can only rate your own complaints');
  if (complaint.status !== 'CLOSED') throw new BadRequestError('You can only rate closed complaints');
  if (!complaint.workerId) throw new BadRequestError('No worker assigned to this complaint');

  const already = await prisma.complaintRating.findFirst({ where: { complaintId: complaint.id } });
  if (already) throw new BadRequestError('You have already rated this complaint');

  const rating = await prisma.complaintRating.create({
    data: {
      complaintId: complaint.id,
      workerId: complaint.workerId,
      studentId: student.id,
      rating: request.rating,
      comment: request.comment,
    },
  });

  const worker = await prisma.user.findUnique({ where: { id: complaint.workerId } });
  return toResponse(rating, worker.name);
}

async function getRatingForComplaint(complaintId) {
  const rating = await prisma.complaintRating.findFirst({ where: { complaintId: BigInt(complaintId) } });
  if (!rating) return null;
  const worker = await prisma.user.findUnique({ where: { id: rating.workerId } });
  return toResponse(rating, worker ? worker.name : 'Unknown');
}

async function getWorkerRatingSummary(workerId) {
  const worker = await prisma.user.findUnique({ where: { id: BigInt(workerId) }, include: { department: true } });
  if (!worker) throw new NotFoundError('Worker not found');

  const agg = await prisma.complaintRating.aggregate({
    where: { workerId: worker.id },
    _avg: { rating: true },
    _count: { rating: true },
  });

  return {
    workerId: worker.id,
    workerName: worker.name,
    department: worker.department ? worker.department.name : null,
    averageRating: agg._avg.rating != null ? Math.round(agg._avg.rating * 10) / 10 : 0.0,
    totalRatings: agg._count.rating || 0,
  };
}

async function getAllWorkerRatings() {
  const workers = await prisma.user.findMany({ where: { role: 'WORKER' } });
  return Promise.all(workers.map((w) => getWorkerRatingSummary(w.id)));
}

module.exports = { submitRating, getRatingForComplaint, getWorkerRatingSummary, getAllWorkerRatings };
